src/coreclr/src/vm/ilmarshalers.cpp

```diff
@@ -139,11 +139,17 @@ void ILDelegateMarshaler::EmitConvertContentsNativeToCLR(ILCodeStream* pslILEmit
     EmitLoadNativeValue(pslILEmit);
     pslILEmit->EmitLDTOKEN(pslILEmit->GetToken(m_pargs->m_pMT));
     pslILEmit->EmitCALL(METHOD__TYPE__GET_TYPE_FROM_HANDLE, 1, 1); // Type System.Type.GetTypeFromHandle(RuntimeTypeHandle handle)
-    pslILEmit->EmitCALL(METHOD__MARSHAL__GET_DELEGATE_FOR_FUNCTION_POINTER, 2, 1); // Delegate System.Marshal.GetDelegateForFunctionPointer(IntPtr p, Type t)
-    EmitStoreManagedValue(pslILEmit);
 
+    // COMPAT: There is a subtle difference between argument and field marshaling with Delegate types.
+    // During field marshaling, the plain Delegate type can be used even though that type doesn't
+    // represent a concrete type. Argument marshaling doesn't permit this so we use the public
+    // API which will validate that Delegate and MulticastDelegate aren't directly used.
```

Comment:
The public API does not actually validate this due to a bug / feature. `Marshal.GetDelegateForFunctionPointer<MulticastDelegate>(...)` will be allowed through and fail in creative way, depending on the runtime version...

Reply:
@jkotas Shoot. That is a good point.... I just realize the implications of that line of code in `Marshal.GetDelegateForFunctionPointer()`.

>will be allowed through and fail in creative way

I think with the new check I added, this will be okay. However, there is that `MethodTable` check in `GetDelegateForFunctionPointerInternal` that assumes non-`null` is that what you mean by "fail in creative way"? If so what do you think about fixing the check in `Marshal.GetDelegateForFunctionPointer()`? At a minimum I will probably add a comment.

Reply:
> fail in creative way

On .NET Framework, it hits AV in the runtime.
On .NET Core 3.1, it failed with MissingMethodException.

I can see that the failure may be more graceful with your latest change. Having a test for this would be useful.

Reply:
See https://github.com/dotnet/runtime/pull/34655